File: python/src/algorithm/sorting/bubble_sort.py
# coding=utf-8
from typing import List


class BubbleSort(object):

    def sort(self, arr: List[int]):
        """
        Time: O(n^2), Space: O(1)
        :param arr:
        :return:
        """
        if not arr:
            return
        for i in range(len(arr) - 1, 0, -1):
            for j in range(i):
                if arr[j] > arr[j + 1]:
                    arr[j], arr[j + 1] = arr[j + 1], arr[i]

    def sort2(self, arr: List[int]):
        """
        Time: O(n^2), Space: O(1)
        :param arr:
        :return:
        """
        
        if not arr:
            return
        for i in range(len(arr) - 1, 0, -1):
            swapped = False
            for j in range(i):
                if arr[j] > arr[j + 1]:
                    arr[j], arr[j + 1] = arr[j + 1], arr[j]
                    swapped = True
            if not swapped:
                return

    def sort3(self, arr: List[int]):
        """
        Time: O(n^2), Space: O(1)
        :param arr:
        :return:
        """
        if not arr:
            return

        i = len(arr) - 1
        while i > 0:
            new_end = 0
            for j in range(i):
                if arr[j] > arr[j + 1]:
                    arr[j], arr[j + 1] = arr[j + 1], arr[j]
                    new_end = i
            i = new_end

File: java/src/main/java/com/ywh/sorting/BubbleSort.java
package com.ywh.sorting;

/**
 * 冒泡排序
 * [排序] [稳定排序]
 *
 * @author ywh
 * @since 13/11/2019
 */
public class BubbleSort {

    /**
     * @param arr
     * @param i
     * @param j
     */
    private void swap(int[] arr, int i, int j) {
        int tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }

    /**
     * end: [n - 1, 0)
     * i: [0, end)
     *
     * Time: O(n^2), Space: O(1)
     *
     * @param arr
     */
    public void sort(int[] arr) {
        if (arr == null || arr.length == 0) {
            return;
        }
        for (int end = arr.length - 1; end > 0; end--) {
            for (int i = 0; i < end; i++) {
                if (arr[i] > arr[i + 1]) {
                    swap(arr, i, i + 1);
                }
            }
        }
    }

    /**
     * 提早返回
     *
     * Time: O(n^2), Space: O(1)
     *
     * @param arr
     */
    public void sortEarlyReturn(int[] arr) {
        if (arr == null || arr.length == 0) {
            return;
        }
        for (int end = arr.length - 1; end > 0; end--) {

            // 如果子循环中没有发生交换，表示排序已完成，直接返回
            boolean swapped = false;
            for (int i = 0; i < end; i++) {
                if (arr[i] > arr[i + 1]) {
                    swap(arr, i, i + 1);
                    swapped = true;
                }
            }
            if (!swapped) {
                return;
            }
        }
    }

    /**
     * 跳过非必要的位置
     *
     * Time: O(n^2), Space: O(1)
     *
     * @param arr
     */
    public void sortSkip(int[] arr) {
        if (arr == null || arr.length == 0) {
            return;
        }
        for (int end = arr.length - 1; end > 0; ) {

            // 记录最后一次发生交换的位置，下次子循环到达这个位置即可
            int newEnd = 0;
            for (int i = 0; i < end; i++) {
                if (arr[i] > arr[i + 1]) {
                    swap(arr, i, i + 1);
                    newEnd = i;
                }
            }
            end = newEnd;
        }

    }

}
